### PR TITLE
[feature-wip](iceberg) Step1: Support create Iceberg external table

### PR DESCRIPTION
**Pull request by qidaye**
## Proposed changes

Close related #7389

Support create Iceberg external table in Doris. 

This is the first step to support Iceberg external table.

### Create Iceberg external table
This pr describes two ways to create Iceberg external tables. Both ways do not require explicitly specifying column definitions, Doris automatically converts them based on Iceberg's column definitions.

1. Create an Iceberg external table directly

```sql
    CREATE [EXTERNAL] TABLE table_name 
    ENGINE = ICEBERG
    [COMMENT "comment"]
    PROPERTIES (
    "iceberg.database" = "iceberg_db_name",
    "iceberg.table" = "icberg_table_name",
    "iceberg.hive.metastore.uris"  =  "thrift://192.168.0.1:9083",
    "iceberg.catalog.type"  =  "HIVE_CATALOG"
    );
```

2. Create an Iceberg database and automatically create all the tables under that db.

```sql
    CREATE DATABASE db_name 
    [COMMENT "comment"]
    PROPERTIES (
    "iceberg.database" = "iceberg_db_name",
    "iceberg.hive.metastore.uris" = "thrift://192.168.0.1:9083",
    "iceberg.catalog.type" = "HIVE_CATALOG"
    );
```

### Show table creation

1. For individual tables you can view them with `help show create table`.

```sql 
mysql> show create table iceberg_db.logs_1;
+--------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table  | Create Table                                                                                                                                                                                                                                                                                                                                                 |
+--------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logs_1 | CREATE TABLE `logs_1` (
  `level` varchar(-1) NOT NULL COMMENT "null",
  `event_time` datetime NOT NULL COMMENT "null",
  `message` varchar(-1) NOT NULL COMMENT "null"
) ENGINE=ICEBERG
COMMENT "ICEBERG"
PROPERTIES (
"iceberg.database" = "doris",
"iceberg.table" = "logs_1",
"iceberg.hive.metastore.uris"  =  "thrift://10.10.10.10:9087",
"iceberg.catalog.type"  =  "HIVE_CATALOG"
) |
+--------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

2. For Iceberg database, you can view it with `help show table creation`.

```sql
mysql> show table creation from iceberg_db;
+--------+---------+---------------------+---------------------------------------------------------+
| Table  | Status  | Create Time         | Error Msg                                               |
+--------+---------+---------------------+---------------------------------------------------------+
| logs   | fail    | 2021-12-14 13:50:10 | Cannot convert unknown type to Doris type: list<string> |
| logs_1 | success | 2021-12-14 13:50:10 |                                                         |
+--------+---------+---------------------+---------------------------------------------------------+
2 rows in set (0.00 sec)
```

  This is a new syntax.
  
  Show table creation records in Iceberg database:
  
  Syntax:
  ```sql
      SHOW TABLE CREATION [FROM db] [LIKE mask]
  ```
## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/catalog/IcebergDatabase.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Map;
+
+/**
+ * Internal represent Iceberg database in Doris
+ */
+public class IcebergDatabase extends Database {
+    private static final Logger LOG = LogManager.getLogger(IcebergDatabase.class);
+
+    private String icebergDb;
+    private Map<String, String> dbProperties = Maps.newHashMap();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/IcebergDatabase.java:40]: I think `dbProperties` can be a common fields of `Database`?
So we don't need to use a derived class.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 2]
```diff
@@ -2643,7 +2653,13 @@ public void createDb(CreateDbStmt stmt) throws DdlException {
                 }
             } else {
                 id = getNextId();
-                Database db = new Database(id, fullDbName);
+                if (engineName.equals("builtin")) {
+                    db = new Database(id, fullDbName);
+                } else if (engineName.equals("iceberg")) {
+                    db = createIcebergDb(stmt, id, icebergTables);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:2659]: `createIcebergDb` may be a time cost methods? but it is executed with lock holding.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 3]
```diff
@@ -4072,6 +4121,62 @@ private void createHiveTable(Database db, CreateTableStmt stmt) throws DdlExcept
...
+     * 1. check table existence in Iceberg
+     * 2. get table schema from Iceberg
+     * 3. convert Iceberg table schema to Doris table schema
+     * 4. create associate table in Doris
+     *
+     * @param db
+     * @param stmt
+     * @throws DdlException
+     */
+    private void createIcebergTable(Database db, CreateTableStmt stmt) throws DdlException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:4136]: Move all iceberg related method to a new class.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 4]
```diff
@@ -2642,16 +2657,24 @@ public void createDb(CreateDbStmt stmt) throws DdlException {
...
         } finally {
             unlock();
         }
         LOG.info("createDb dbName = " + fullDbName + ", id = " + id);
+
+        // create tables in iceberg database
+        if (db.getDbProperties().getIcebergProperty().isExist()) {
+            IcebergProperty icebergProperty = db.getDbProperties().getIcebergProperty();
+            IcebergCatalog icebergCatalog = IcebergCatalogMgr.getCatalog(icebergProperty);
+            List<TableIdentifier> icebergTables = icebergCatalog.listTables(icebergProperty.getDatabase());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:2672]: What if the `listTables` throw exception?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 5]
```diff
@@ -4615,6 +4656,9 @@ public boolean unprotectDropTable(Database db, Table table, boolean isForceDrop,
             // drop all temp partitions of this table, so that there is no temp partitions in recycle bin,
             // which make things easier.
             ((OlapTable) table).dropAllTempPartitions();
+        } else if (table.getType() == TableType.ICEBERG) {
+            // drop Iceberg database table creation record
+            icebergTableCreationRecordMgr.deRegisterTable(db, (IcebergTable) table);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:4661]:
```suggestion
            icebergTableCreationRecordMgr.deregisterTable(db, (IcebergTable) table);
```

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            dbToTableIdentifiers.get(db).put(identifier, icebergProperty);
+        } else {
+            Map<TableIdentifier, IcebergProperty> identifierToProperties = Maps.newConcurrentMap();
+            identifierToProperties.put(identifier, icebergProperty);
+            dbToTableIdentifiers.put(db, identifierToProperties);
+        }
+        LOG.info("Register a new table[{}] to database[{}]", identifier.name(), db.getFullName());
+    }
+
+    public void deRegisterDb(Database db) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java:84]:
```suggestion
    public void deregisterDb(Database db) {
```

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                            prop.writeTimeFormat(new Date(System.currentTimeMillis())), e.getMessage());
+                    LOG.warn("Failed create table[{}], error: {}", identifier.name(), e.getMessage());
+                }
+            }
+        }
+        removeDuplicateTables();
+    }
+
+    private void addTableCreationRecord(String db, String table, String status, String createTime, String errorMsg) {
+        writeLock();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java:152]: Use `try...finally` to wrap the lock

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 4 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:2674]: not resolved?

File: fe/fe-core/src/main/java/org/apache/doris/common/FeMetaVersion.java [hunk 8]
```diff
@@ -222,6 +222,8 @@
     public static final int VERSION_104 = 104;
     // change replica to replica allocation
     public static final int VERSION_105 = 105;
+    // add iceberg database
+    public static final int VERSION_106 = 106;
     // note: when increment meta version, should assign the latest version to VERSION_CURRENT
-    public static final int VERSION_CURRENT = VERSION_105;
+    public static final int VERSION_CURRENT = VERSION_106;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/common/FeMetaVersion.java:228]: Remove this

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        if (dbToTableIdentifiers.containsKey(db)) {
+            TableIdentifier identifier = TableIdentifier.of(table.getIcebergDb(), table.getIcebergTbl());
+            Map<TableIdentifier, IcebergProperty> identifierToProperties = dbToTableIdentifiers.get(db);
+            identifierToProperties.remove(identifier);
+        }
+        if (dbToTableToCreationRecord.containsKey(db.getFullName())) {
+            Map<String, IcebergTableCreationRecord> recordMap = dbToTableToCreationRecord.get(db.getFullName());
+            recordMap.remove(table.getName());
+        }
+        LOG.info("DeRegister table[{}] from database[{}]", table.getName(), db.getFullName());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java:100]:
```suggestion
        LOG.info("Deregister table[{}] from database[{}]", table.getName(), db.getFullName());
```

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+            dbToTableIdentifiers.get(db).put(identifier, icebergProperty);
+        } else {
+            Map<TableIdentifier, IcebergProperty> identifierToProperties = Maps.newConcurrentMap();
+            identifierToProperties.put(identifier, icebergProperty);
+            dbToTableIdentifiers.put(db, identifierToProperties);
+        }
+        LOG.info("Register a new table[{}] to database[{}]", identifier.name(), db.getFullName());
+    }
+
+    public void deregisterdb(Database db) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java:84]:
```suggestion
    public void deregisterDb(Database db) {
```

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/SHOW TABLE CREATION.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    语法：
+        SHOW TABLE CREATION [FROM db_name] [LIKE table_name_wild];
+        
+    说明：
+        1) 如果不指定 db_name，使用当前默认 db
+        2) 如果使用 LIKE，则会匹配表名中包含 table_name_wild 的建表任务
+
+## example
+
+    1. 展示默认 Iceberg db 中所有的建表任务
```
Comment (morningman) [inline docs/zh-CN/sql-reference/sql-statements/Data Manipulation/SHOW TABLE CREATION.md:41]: Add an result example to explain the show result

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 4 above)
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:2674]: Added a new `DorisIcebergException` which is a `RuntimeException` in `IcebergCatalog.listTable`.

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/SHOW TABLE CREATION.md (hunk 11 above)
Comment (qidaye) [reply docs/zh-CN/sql-reference/sql-statements/Data Manipulation/SHOW TABLE CREATION.md:47]: Done

File: fe/fe-core/src/main/java/org/apache/doris/common/FeMetaVersion.java (hunk 8 above)
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/common/FeMetaVersion.java]: Done

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java (hunk 9 above)
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java]: Done

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java (hunk 10 above)
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergTableCreationRecordMgr.java]: Done

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 12]
```diff
@@ -2631,7 +2636,15 @@ public Frontend getFeByName(String name) {
...
         final String clusterName = stmt.getClusterName();
         String fullDbName = stmt.getFullDbName();
-        long id = 0L;
+        Map<String, String> properties = stmt.getProperties();
+
+        long id = getNextId();
+        Database db = new Database(id, fullDbName);
+        db.setClusterName(clusterName);
+        // check and analyze database properties before create database
+        db.getDbProperties().putAll(properties);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:2645]: merge these 2 methods:
```
db.getDbProperties().putAll(properties);
db.getDbProperties().checkAndBuildProperties();
```
to
```
db.addAndBuildProperties(properties);
```

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**